### PR TITLE
Apache BEAM SDK updates to handle vector length column annotation

### PR DESCRIPTION
**Pull request by ambujojha**
1. Updates in SpannerSchema.java
2. No updates necessary in MutationSizeEstimator.java

### CODE REVIEW

File: v1/src/test/java/com/google/cloud/teleport/spanner/SpannerServerResource.java [hunk 1]
```diff
@@ -41,7 +41,8 @@ public class SpannerServerResource extends ExternalResource {
   private static final String EMULATOR_HOST = System.getenv("SPANNER_EMULATOR_HOST");
   private static final String DEFAULT_PROJECT_ID = "span-cloud-testing";
   private static final String DEFAULT_INSTANCE_ID = "test-instance";
-  private final String host = "https://spanner.googleapis.com";
+  // private final String host = "https://spanner.googleapis.com";
+  private final String host = "https://staging-wrenchworks.sandbox.googleapis.com/";
```
Comment (darshan-sj) [inline v1/src/test/java/com/google/cloud/teleport/spanner/SpannerServerResource.java:45]: This should be changed back.

File: v1/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java [hunk 2]
```diff
@@ -148,6 +150,14 @@ static KeyPart create(String field, boolean desc) {
   @AutoValue
   abstract static class Column implements Serializable {
 
+    private static final Pattern EMBEDDING_VECTOR_PATTERN =
+        Pattern.compile(
+            "^ARRAY<([a-zA-Z0-9]+)>\\(vector_length=>\\d+\\)$", Pattern.CASE_INSENSITIVE);
+
+    private static final Pattern PG_EMBEDDING_VECTOR_PATTERN =
+        Pattern.compile("^(\\D+)\\[\\]\\svector\\slength\\s\\d+$", Pattern.CASE_INSENSITIVE);
```
Comment (darshan-sj) [inline v1/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java:158]: I thought erasable type (vector_length=>) is only for GoogleSQL and in Postgres dialect it is implemented as column option. Is it correct?

Comment (ambujojha) [reply v1/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java:158]: No. its an erasable type in both googleSQL and Postgres.

**Review by darshan-sj — APPROVED** [review]

File: v1/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java [hunk 3]
```diff
@@ -192,13 +202,30 @@ private static Type parseSpannerType(String spannerType, Dialect dialect) {
...
+            // Substring "ARRAY<xxx>" or substring "ARRAY<xxx>(vector_length=>yyy)"
+
+            String arrayElementType;
+            // Handle vector_length annotation
+            Matcher m = EMBEDDING_VECTOR_PATTERN.matcher(spannerType);
+            if (m.find()) {
+              arrayElementType = m.group(1);
+            }
+            else {
+              arrayElementType = spannerType.substring(6, spannerType.length() - 1);
```
Comment (jonathan-lemos) [inline v1/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java:214]: Nit: don't like substring here for the same reason as last PR, but I don't consider it to be a blocking issue.

**Review by jonathan-lemos — APPROVED** [review]

File: v1/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java (hunk 3 above)
Comment (ambujojha) [reply v1/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java:214]: This is already existing code and beyond the scope of this change IMO.
The code that handles vector length uses regex patten matching.

**Review by darshan-sj — APPROVED** [review]

**Comment by Abacn** [comment]
Please sync this update to Beam repo. LocalSpannerIO will be removed and this change will get lost

**Comment by ambujojha** [comment]
> Please sync this update to Beam repo. LocalSpannerIO will be removed (#1362) and this change will get lost

Sure. How much time do I have?

**Comment by Abacn** [comment]
> > Please sync this update to Beam repo. LocalSpannerIO will be removed (#1362) and this change will get lost
> 
> Sure. How much time do I have?

The next Beam release cut date is Apr 17, 2024. It's nice to have Beam PR merge it by that time so this would unblock tearing down LocalSpannerIO when Templates upgrade to Beam 2.56.0

**Pull request merged**